core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -251,6 +255,9 @@
     private final TypeProvider symbolTypes;
     private final boolean isDescribe;
 
+    // Cache from SQL type name to Type; every Type in the cache has a CAST defined from VARCHAR
+    private final Cache<String, Type> varcharCastableTypeCache = buildNonEvictableCache(CacheBuilder.newBuilder().maximumSize(1000));
```

Comment:
does it need to be a cache since `ExpressionAnalyzer` are not long lived and we keep both `name` (within `Literal`) and `Type` within `expressionTypes`. Thus no savings in mem

Reply:
Good call.

There actually will be 6 entries here, because there are only 6 types for which `LiteralEncoder` produces a `GenericLiteral` and none of them is parametric. 

Using an unbounded map is OK for me. Using a cache is more defensive programming, and conveys the meaning: this, after all, serves as a cache.

Reply:
@findepi `GenericLiteral#type` is a "raw" string as typed by the user. If you want to have at most 6 entries, you need to canonicalize it. Same applies to the cache in `LiteralInterpreter`.

Reply:
I think about 6 entries because practically all expressions are normalized, so they are constants re-encoded into Expression via `LiteralEncoder`.